### ISSUE TITLE
ENH: speeding up pairplot and alike

### ISSUE THREAD

**Issue by tupui** [mwaskom/seaborn#2609]
I think there might be room for speeding up methods like `sns.pairplot`.

I want to be able to draw huge plots (100x100 or more) and this is of course unrealistic with the current code. I wonder if we could parallelise some of the loops like this one: 

https://github.com/mwaskom/seaborn/blob/445a54aa46e279406949e0a3c2eed88d6cf80223/seaborn/axisgrid.py#L649

On matplotlib's side I am confident there would be no blocker as they have this tutorial https://matplotlib.org/stable/gallery/misc/multiprocess_sgskip.html 

I am not entirely sure if this would be the correct place to do it. Let me know if this is of interest/feasible and I would be happy to help.

**Comment by mwaskom** [comment]
Before engaging in this, there would need to be a pretty serious effort to identify the actual bottlenecks. In my experience, one of the slowest parts of making a very large `PairGrid` is in the figure initialization (i.e. entirely in matplotlib; this is a [known limitation](https://github.com/matplotlib/matplotlib/issues/6664)). It's likely that parallelizing seaborn operations that add artists to the figure could save time (especially for more compute-heavy functions like `kdeplot`), but how much time? 80%? 50%? 10%?

I just did an extremely simple test, and for a 10x10 grid, initializing the figure takes ~2x the time of mapping `scatterplot`, so if you spread that across, say, 4 cores and ignore overhead, you're only talking about a 25% improvement.

Cost-benefit analysis is important because building in parallelization would add complexity and decrease robustness. In some cases it would not be possible. For one, if you are mapping a matplotlib function or a custom function, it makes use of the pyplot global state, so it's not possible to parallelize. I'm also not sure that example is a particularly good guide to what you want to do. It seems to be about initializing one figure in a separate process and piping data over to it, like for a dashboard.

Parallelizing `PairGrid` would be different because you would be modifying axes in parallel that all live in the same figure. You'd need to find some way to share memory across the processes so that they are actually all modifying the same object (rather than serializing the figure and then sending multiple instances of it to different processes). In the example, the figure always lives in the follower process. And then you'd have to contend with problems arising from interactions between the processes (i.e. PairGrid figures share axes by defaults, so what happens when two parallel processes set different axis limits?).

I'm also not all that convinced that the goal of drawing a 100x100 pairplot is particularly important to enable, because I'm skeptical that it's possible to make an informative figure at that scale. That said it's still pretty slow even for more reasonably large figures (i.e. ~10-20 variables) so I think speeding it up would be great, but again, the major bottleneck is on the matplotlib side. I'd say it would be better to spend development effort improving the efficiency of matplotlib figure initialization (especially since that will have a much larger audience than users of `seaborn.PairGrid`).

Also, if you really want to make *HUGE PLOTS*, I think you may want to look into some of the more performant opengl-based visualization libraries, like vispy.

So I'm not totally opposed, but I would need to be convinced it's both feasible and worth it with an argument beyond "hey this is slow and, look, a serial loop".

**Comment by tupui** [comment]
Thanks for the fast reply!

> Before engaging in this, there would need to be a pretty serious effort to identify the actual bottlenecks. In my experience, one of the slowest parts of making a very large `PairGrid` is in the figure initialization (i.e. entirely in matplotlib; this is a [known limitation](https://github.com/matplotlib/matplotlib/issues/6664)). It's likely that parallelizing seaborn operations that add artists to the figure could save time (especially for more compute-heavy functions like `kdeplot`), but how much time? 80%? 50%? 10%?

I was not aware of this, thanks for pointing it out. I did not profile the function to see exactly what was happening, I will see if I can do that.

> I'm also not all that convinced that the goal of drawing a 100x100 pairplot is particularly important to enable, because I'm skeptical that it's possible to make an informative figure at that scale.

For my use case that would actually still work. For the context, I want to do a visual analysis of random samples (using QMC methods):

```python
# with latest SciPy (1.7)
from scipy.stats import qmc
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

engine = qmc.Sobol(d=25, scramble=True)  # I want to go up with d here
sample = engine.random(512)
sns.pairplot(pd.DataFrame(sample), diag_kind="hist", diag_kws={"bins": 16},
             # corner=True
)
```

Being able to do this allows to quickly see if we have patterns on some group of 2-2 dimensions. Arguably, there are other ways of doing this. Like I could do a parallel loop myself on all the combinations and then merge all figures into a large PDF or a movie. Or use a metric like the discrepancy or other criteria and hope to detect these. I would need to experiment here.

<img width="768" alt="Screenshot 2021-06-29 at 15 16 49" src="https://user-images.githubusercontent.com/23188539/123804833-d6f28980-d8ed-11eb-805a-563eca91a84f.png">

> Also, if you really want to make _HUGE PLOTS_, I think you may want to look into some of the more performant opengl-based visualization libraries, like vispy.

I will have a look, thank you for the suggestion.

**Comment by mwaskom** [comment]
Yeah I do see how some patterns pop out here, but I think I would start with e.g. computing a pairwise mutual information matrix and making a heatmap. Or a [heat-scatter](http://seaborn.pydata.org/examples/heat_scatter.html), either dual-encoding the MI measure or representing two potentially informative measures with hue and size independently.

**Comment by tupui** [comment]
Thanks for the link to the heat-scatter. I will play with that and some measures.

In the meantime, feel free to close the issue if you think this would be more something to address on matplotlib's side. Or that this is not realistic/out of scope here.

**Comment by mwaskom** [comment]
Here's an extraordinarily simple proof of principle that I threw together just for exploration:

```python
import numpy as np
import matplotlib.pyplot as plt
from joblib.parallel import Parallel, delayed

def pairplot(ax, x, y):
    ax.scatter(x, y)
    
xs = np.random.randn(4, 10000)
ys = np.random.randn(4, 10000)

f, axs = plt.subplots(1, 4, sharex=True, sharey=True)
Parallel(n_jobs=4)(delayed(pairplot)(ax, x, y) for ax, x, y in zip(axs, xs, ys))
```

With `n_jobs=1` this works as expected but with `n_jobs=4` I get four blank axes. I've done no further debugging or tuning, however.

I think I'll close for now, not because I'm opposed to the idea of parallelizing the axisgrid objects in principle, but because I'm skeptical that it's *possible* for reasons outside of seaborn's control. I'd rather consider the usefulness/feasibility in the context of an existing prototype of how one might parallelize operations over the axes in a single matplotlib figure in general.
